pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/model/Environment.groovy

```diff
@@ -65,31 +65,86 @@ public class Environment implements Serializable {
+            String val = v.value.toString()
+            if (!v.isLiteral) {
+                val = replaceEnvDotInCurlies(val)
+            }
+            if (v.isLiteral || (v.value.toString().startsWith('$'))) {
+                [(k): val]
+            } else {
+                [(k): Utils.trimQuotes(val)]
+            }
+        }
 
-        EnvVars contextEnv = new EnvVars((Map<String,String>)script.getProperty("params"))
+        Map<String,String> alreadySet = new TreeMap<>()
+        alreadySet.putAll(CpsThread.current()?.getExecution()?.getShell()?.getContext()?.variables?.findAll { k, v ->
+            k instanceof String && v instanceof String
+        })
         if (withContext) {
-            contextEnv.overrideExpandingAll(((EnvActionImpl)script.getProperty("env")).getEnvironment())
+            alreadySet.putAll(((EnvActionImpl)script.getProperty("env")).getEnvironment())
         }
-
-        newEnv = new EnvVars(contextEnv)
-
+        alreadySet.putAll((Map<String,String>)script.getProperty("params"))
         if (parent != null) {
-            newEnv.overrideExpandingAll(parent.resolveEnvVars(script, false))
+            alreadySet.putAll(parent.resolveEnvVars(script, false))
         }
 
-        Map<String,String> overrides = getMap().collectEntries { k, v ->
-            if (v.isLiteral || (v.value.toString().startsWith('$'))) {
-                [(k): v.value.toString()]
-            } else {
-                [(k): Utils.trimQuotes(v.value.toString())]
+        List<String> unsetKeys = []
+        unsetKeys.addAll(overrides.keySet())
+
+        GroovyShell shell = new GroovyShell(new Binding(alreadySet))
+        shell.setProperty("params", (Map<String,String>)script.getProperty("params"))
+
+        int unsuccessfulCount = 0
+        while (!unsetKeys.isEmpty() && unsuccessfulCount <= overrides.size()) {
+            String nextKey = unsetKeys.first()
+
+            unsetKeys.remove(nextKey)
+            try {
+                String resolved = overrides.get(nextKey)
+                // Only do the eval here if the string actually contains another environment variable we know about.
+                // This is to defer evaluation of things like ${WORKSPACE} or currentBuild.getNumber() until later.
+                if (containsVariable(resolved, overrides.keySet()) ||
+                    containsVariable(resolved, alreadySet.keySet())) {
+                    resolved = shell.evaluate(Utils.prepareForEvalToString(resolved))
+                }
+
+                alreadySet.put(nextKey, resolved)
+                shell.setVariable(nextKey, resolved)
+                unsuccessfulCount = 0
+            } catch (_) {
+                unsuccessfulCount++
+                unsetKeys.add(nextKey)
+            }
+        }
+
+        return alreadySet
+    }
+
+    private boolean containsVariable(String var, Set<String> keys) {
+        def group = (var =~ /(\$\{.*?\})/)
```

Comment:
Iirc Environment.groovy isn't executed in the CPS engine so a regex should be safe here.